drivers/i2c/i2c_cc32xx.c

```diff
@@ -0,0 +1,431 @@
+			 */
+			MAP_I2CMasterControl(base,
+				       I2C_MASTER_CMD_BURST_RECEIVE_CONT_NACK);
+		}
+	} else {
+		/* Next state: Idle mode */
+		data->state = I2C_CC32XX_IDLE_MODE;
+
+		/*
+		 * No more data needs to be received, so follow up with a
+		 * STOP bit
+		 * Again, there is no equivalent macro (0x00000004) so
+		 * I2C_MASTER_CMD_BURST_RECEIVE_STOP is used
+		 */
+		MAP_I2CMasterControl(base,
+				     I2C_MASTER_CMD_BURST_RECEIVE_STOP);
+	}
+}
+
+static void i2c_cc32xx_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	u32_t base = DEV_BASE(dev);
+	struct i2c_cc32xx_data *data = DEV_DATA(dev);
+	u32_t err_status;
+
+	/* Get the interrupt status of the I2C controller */
+	err_status = MAP_I2CMasterErr(base);
+
+	/* Clear interrupt source to avoid additional interrupts */
+	MAP_I2CMasterIntClear(base);
+
+	/* Check for I2C Errors */
+	if ((err_status == I2C_MASTER_ERR_NONE) ||
+	    (data->state == I2C_CC32XX_ERROR)) {
+		/* No errors, now check what we need to do next */
+		switch (data->state) {
+		/*
+		 * ERROR case is OK because if an Error is detected, a STOP bit
+		 * is sent; which in turn will call another interrupt. This
+		 * interrupt call will then post the transfer_complete semaphore
+		 * to unblock the transfer function.
+		 */
+		case I2C_CC32XX_ERROR:
+		case I2C_CC32XX_IDLE_MODE:
+			/* Indicate transfer complete */
+			k_sem_give(&data->transfer_complete);
+			break;
+
+		case I2C_CC32XX_WRITE_MODE:
+			_i2c_cc32xx_isr_handle_write(base, data);
+			break;
+
+		case I2C_CC32XX_READ_MODE:
+			_i2c_cc32xx_isr_handle_read(base, data);
+			break;
+
+		default:
+			data->state = I2C_CC32XX_ERROR;
+			break;
+		}
+	} else {
+		/* Some sort of error happened */
+		data->state = I2C_CC32XX_ERROR;
+
+		if (err_status & (I2C_MASTER_ERR_ARB_LOST |
+				  I2C_MASTER_ERR_ADDR_ACK)) {
+			k_sem_give(&data->transfer_complete);
+		} else {
+			/* Send a STOP bit to end I2C communications */
+			/*
+			 * I2C_MASTER_CMD_BURST_SEND_ERROR_STOP -and-
+			 * I2C_MASTER_CMD_BURST_RECEIVE_ERROR_STOP
+			 * have the same values
+			 */
+			MAP_I2CMasterControl(base,
+					  I2C_MASTER_CMD_BURST_SEND_ERROR_STOP);
+			k_sem_give(&data->transfer_complete);
+		}
+	}
+}
+
+static int i2c_cc32xx_init(struct device *dev)
+{
+	u32_t base = DEV_BASE(dev);
+	const struct i2c_cc32xx_config *config = DEV_CFG(dev);
+	struct i2c_cc32xx_data *data = DEV_DATA(dev);
+	u32_t bitrate_cfg;
+	int error;
+	u32_t regval;
+
+	k_sem_init(&data->mutex, 1, UINT_MAX);
+	k_sem_init(&data->transfer_complete, 0, UINT_MAX);
+
+	/* In case of app restart: disable I2C module, clear NVIC interrupt */
+	/* Note: this was done *during* pinmux setup in SimpleLink SDK. */
+	MAP_I2CMasterDisable(base);
+	/* Clear INT_I2CA0 */
+	MAP_IntPendClear((unsigned long)(TI_CC32XX_I2C_40020000_IRQ_0 + 16));
+
+	configure_i2c_irq();
+
+	data->state = I2C_CC32XX_IDLE_MODE;
+
+	/* Take I2C hardware semaphore. */
+	regval = HWREG(0x400F7000);
```

Comment:
can we cleanup this magic numbers at all?

Reply:
yup.